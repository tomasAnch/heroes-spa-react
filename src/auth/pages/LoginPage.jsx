import { useContext } from "react";
import { useNavigate } from "react-router-dom";
import { AuthContext } from "../context";

export const LoginPage = () => {

  const { login } = useContext( AuthContext )
  const navigate = useNavigate();

  const onLogin = () => {

    const lastPath = localStorage.getItem('lastPath') || '/'

    login( 'Tomas Anchorena' )

    navigate( lastPath, {
      replace: true,
    });
  };

  return (
    <section className="relative flex flex-wrap lg:h-screen lg:items-center bg-gray-700">
      <div className="w-full px-4 py-12 sm:px-6 sm:py-16 lg:w-1/2 lg:px-8 lg:py-24">
        <div className="mx-auto max-w-lg text-center font-roboto">
          <h1 className="text-2xl font-bold sm:text-3xl text-gray-300">
            Login to see your favorite Heroes!
          </h1>
          <button
            className="group relative inline-block focus:outline-none focus:ring my-12"
            onClick={ onLogin }
          >
            <span className="absolute inset-0 translate-x-0 translate-y-0 bg-gray-300 transition-transform group-hover:translate-y-1.5 group-hover:translate-x-1.5"></span>

            <span className="text-black relative inline-block border-2 border-current px-8 py-3 text-sm font-bold uppercase tracking-widest">
              Login
            </span>
          </button>
        </div>
      </div>

      <div className="relative h-64 w-full sm:h-96 lg:h-full lg:w-1/2">
        <img
          alt="Heroes Login"
          src="https://wallpapercave.com/wp/G5HNZDV.jpg"
          className="absolute inset-0 h-full w-full object-cover"
        />
      </div>
    </section>
  );
};
