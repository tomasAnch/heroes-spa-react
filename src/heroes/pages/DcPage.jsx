import { HeroList } from "../components/HeroList";

export const DcPage = () => {
  return (
    <>
      <div className="py-10 mx-auto font-poppins">
        <div className="flex justify-center items-center">
          <h1 className="pb-10 text-5xl font-bold text-white">
            DC Characters
          </h1>
        </div>
        <HeroList publisher="DC Comics" />
      </div>
    </>
  );
};
