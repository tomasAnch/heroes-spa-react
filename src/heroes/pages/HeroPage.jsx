import { Navigate, useNavigate, useParams } from "react-router-dom";
import { getHeroById } from "../helpers";
import { useMemo } from "react";

export const HeroPage = () => {
  const { id } = useParams();
  const navigate = useNavigate()

  const hero = useMemo( () => getHeroById( id ), [ id ] )

  const onNavigateBack = () => {
    navigate(-1)
  }
 
  if (!hero) {
    return <Navigate to="/marvel" />;
  }

  return (
    <div className="grid grid-cols-2 gap-14 py-12 animate__animated animate__backInUp">
      <div className="col-span-1">
        <img
          src={`/src/assets/heroes/${id}.jpg`}
          alt={hero.superhero}
          className="w-full"
        />
      </div>

      <div className="col-span-1 font-roboto text-white">
        <h3 className="text-center py-6 text-5xl font-semibold">
          {hero.superhero}
        </h3>

        <dl className="max-w-md divide-y text-white divide-gray-700">
          <div className="flex flex-col pb-3">
            <dt className="mb-1 md:text-lg text-gray-400">
              Alter Ego:
            </dt>
            <dd className="text-lg font-semibold">{hero.alter_ego}</dd>
          </div>
          <div className="flex flex-col py-3">
            <dt className="mb-1 md:text-lg text-gray-400">
              Publisher:
            </dt>
            <dd className="text-lg font-semibold">{hero.publisher}</dd>
          </div>
          <div className="flex flex-col pt-3">
            <dt className="mb-1 md:text-lg text-gray-400">
              First Appearance:
            </dt>
            <dd className="text-lg font-semibold">{hero.first_appearance}</dd>
          </div>
        </dl>

        <h5 className="mt-10 text-2xl">Characters</h5>
        <p className="mt-2">{hero.characters}</p>

        <button className="relative inline-flex items-center justify-center p-0.5 my-4 -mx-0.5 overflow-hidden text-sm font-medium  rounded-lg group bg-gradient-to-br from-green-400 to-blue-600 group-hover:from-green-400 group-hover:to-blue-600 hover:text-white text-white focus:ring-4 focus:outline-none focus:ring-green-200 ring-green-800" onClick={ onNavigateBack }>
          <span className="relative px-5 py-2.5 transition-all ease-in duration-75 bg-gray-900 rounded-md group-hover:bg-opacity-0">
            Regresar
          </span>
        </button>
      </div>
    </div>
  );
};
