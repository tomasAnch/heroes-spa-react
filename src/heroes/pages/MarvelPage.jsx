import { HeroList } from "../components/HeroList"

export const MarvelPage = () => {
  return(
    <>
      <div className="py-10 mx-auto font-poppins">
        <div className="flex justify-center items-center">
          <h1 className="pb-10 text-5xl font-bold text-white">Marvel Characters</h1>
        </div>
        <HeroList publisher='Marvel Comics' />
      </div>
    </>
  )
}