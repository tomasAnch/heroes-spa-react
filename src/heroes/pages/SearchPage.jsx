import { useLocation, useNavigate } from "react-router-dom";
import { useForm } from "../../hooks/useForm";
import queryString from "query-string"

import { getHeroesByName } from "../helpers/getHeroesByName";

import { HeroCard } from '../components/HeroCard'

export const SearchPage = () => {

  const navigate = useNavigate()
  const location = useLocation()

  const { q = '' } = queryString.parse( location.search )
  const heroes = getHeroesByName( q )

  const showSearch = (q.length === 0)
  const showError = (q.length > 0) && heroes.length === 0

  const { searchText, onInputChange } = useForm({
    searchText: q
  })

  const onSearchSubmit = (e) => {
    e.preventDefault()
    // if(searchText.trim().length <= 1) return;

    navigate(`?q=${ searchText }`)
  }

  return (
    <div className="grid grid-cols-2 gap-14 justify-center items-center place-content-center mx-auto text-white">
      <div className="col-span-1 py-12">
        <h1 className="font-poppins font-bold text-center text-3xl">Search a Hero!</h1>

        <form className="pt-12" onSubmit={ onSearchSubmit } aria-label="form">
          <label
            htmlFor="search"
            className="mb-2 text-sm font-medium sr-only text-white"
          >
            Search
          </label>
          <div className="relative">
            <div className="absolute inset-y-0 left-0 flex items-center pl-3 pointer-events-none">
              <svg
                className="w-4 h-4 text-gray-400"
                aria-hidden="true"
                xmlns="http://www.w3.org/2000/svg"
                fill="none"
                viewBox="0 0 20 20"
              >
                <path
                  stroke="currentColor"
                  strokeLinecap="round"
                  strokeLinejoin="round"
                  strokeWidth="2"
                  d="m19 19-4-4m0-7A7 7 0 1 1 1 8a7 7 0 0 1 14 0Z"
                />
              </svg>
            </div>
            <input
              type="text"
              placeholder="In which hero are you in?"
              className="block w-72 px-6 py-4 pl-10 text-sm border rounded-lg bg-gray-700 border-gray-600 placeholder-gray-400 text-white focus:ring-blue-500 focus:border-blue-500"
              name="searchText"
              autoComplete="off"
              value={ searchText }
              onChange={ onInputChange }
            />
            <button
              type="submit"
              className="text-white absolute right-2.5 bottom-2.5 focus:ring-4 focus:outline-none font-medium rounded-lg text-sm px-4 py-2 bg-blue-900 hover:bg-blue-700 focus:ring-blue-800"
            >
              Search
            </button>
          </div>
        </form>
      </div>
      <div className="col-span-1 py-12 -mt-2">
        <h1 className="font-poppins font-bold text-center text-3xl">Results</h1>
        
        <div className="pt-12">
          <div className="px-6 bg-gray-700 py-4 mx-auto -mb-1.5 border rounded-lg" style={{ display: showSearch ? '' : 'none' }}>
            <p>Hero</p>
          </div>
          <div 
            aria-label="alert-error"
            className="px-6 bg-red-400 py-4 mx-auto my-4 border rounded-lg" 
            style={{ display: showError ? '' : 'none' }}
          >
            <p className="text-red-900">There's no result for { q }</p>
          </div>

          {
            heroes.map( (hero) => (
              <div className="h-full">
                <HeroCard key={hero.id} {...hero} />
              </div>
            ))
          }

        </div>
      </div>
    </div>
  );
};
