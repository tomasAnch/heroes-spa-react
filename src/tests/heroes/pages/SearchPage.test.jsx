import { fireEvent, render, screen } from "@testing-library/react"
import { MemoryRouter } from "react-router-dom"
import { SearchPage } from "../../../heroes/pages/SearchPage"

const mockedUseNavigate = jest.fn()

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useNavigate: () => mockedUseNavigate,
}))

describe('Pruebas en <SearchPage />', () => {

  beforeEach( () => jest.clearAllMocks() )
  
  test('Debe mostrarse correctamente con valores por defecto', () => {
    
    const { container } = render (
      <MemoryRouter>
        <SearchPage />
      </MemoryRouter>
    )

    expect( container ).toMatchSnapshot()

  })
  
  test('Debe mostrar a Batman y el input con el valor del queryString', () => {
    
    render (
      <MemoryRouter initialEntries={['/search?q=batman']}>
        <SearchPage />
      </MemoryRouter>
    )

    const input = screen.getByRole('textbox')
    expect( input.value ).toBe('batman')

    const img = screen.getByRole('img')
    expect( img.src ).toContain('/assets/heroes/dc-batman.jpg')

    const alertError = screen.getByLabelText('alert-error')
    expect( alertError.style.display ).toBe('none')

  })

  test('Debe mostrar un error si no se encuentra el hero', () => {
    
    render(
      <MemoryRouter initialEntries={['/search?q=batman123']}>
        <SearchPage />
      </MemoryRouter>
    )

    const alertError = screen.getByLabelText('alert-error')
    expect( alertError.style.display ).toBe('')

  })

  test('Debe llamar el navigate a la pantalla nueva', () => {
    
    render(
      <MemoryRouter initialEntries={['/search']}>
        <SearchPage />
      </MemoryRouter>
    )

    const input = screen.getByRole('textbox')
    fireEvent.change( input, { target: { name: 'searchText', value: 'superman' } } )

    const form = screen.getByRole('form')
    fireEvent.submit( form )

    expect( mockedUseNavigate ).toHaveBeenCalledWith('?q=superman')
  })
  
  
  

})
