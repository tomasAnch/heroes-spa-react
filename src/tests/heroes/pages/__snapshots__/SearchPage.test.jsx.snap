// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <SearchPage /> Debe mostrarse correctamente con valores por defecto 1`] = `
<div>
  <div
    class="grid grid-cols-2 gap-14 justify-center items-center place-content-center mx-auto text-white"
  >
    <div
      class="col-span-1 py-12"
    >
      <h1
        class="font-poppins font-bold text-center text-3xl"
      >
        Search a Hero!
      </h1>
      <form
        aria-label="form"
        class="pt-12"
      >
        <label
          class="mb-2 text-sm font-medium sr-only text-white"
          for="search"
        >
          Search
        </label>
        <div
          class="relative"
        >
          <div
            class="absolute inset-y-0 left-0 flex items-center pl-3 pointer-events-none"
          >
            <svg
              aria-hidden="true"
              class="w-4 h-4 text-gray-400"
              fill="none"
              viewBox="0 0 20 20"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="m19 19-4-4m0-7A7 7 0 1 1 1 8a7 7 0 0 1 14 0Z"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
              />
            </svg>
          </div>
          <input
            autocomplete="off"
            class="block w-72 px-6 py-4 pl-10 text-sm border rounded-lg bg-gray-700 border-gray-600 placeholder-gray-400 text-white focus:ring-blue-500 focus:border-blue-500"
            name="searchText"
            placeholder="In which hero are you in?"
            type="text"
            value=""
          />
          <button
            class="text-white absolute right-2.5 bottom-2.5 focus:ring-4 focus:outline-none font-medium rounded-lg text-sm px-4 py-2 bg-blue-900 hover:bg-blue-700 focus:ring-blue-800"
            type="submit"
          >
            Search
          </button>
        </div>
      </form>
    </div>
    <div
      class="col-span-1 py-12 -mt-2"
    >
      <h1
        class="font-poppins font-bold text-center text-3xl"
      >
        Results
      </h1>
      <div
        class="pt-12"
      >
        <div
          class="px-6 bg-gray-700 py-4 mx-auto -mb-1.5 border rounded-lg"
        >
          <p>
            Hero
          </p>
        </div>
        <div
          aria-label="alert-error"
          class="px-6 bg-red-400 py-4 mx-auto my-4 border rounded-lg"
          style="display: none;"
        >
          <p
            class="text-red-900"
          >
            There's no result for 
          </p>
        </div>
      </div>
    </div>
  </div>
</div>
`;
